From 472838cfa6c5107e70e0481e88a9ea501d031686 Mon Sep 17 00:00:00 2001
From: Chirag Parekh <chirag.parekh@xilinx.com>
Date: Wed, 8 Feb 2017 03:08:50 -0800
Subject: [PATCH 1295/1566] pinctrl: zynqmp: Resolved pin conflicts

commit  5b41a9ca7435338a46da8c0a104bc6b87b74e13c from
https://github.com/Xilinx/linux-xlnx.git

Separated groups for NAND chip enable, data strobe and ready/busy
since common group was resulting in conflict with CAN for
zynqmp zc1751-xm016 platform. Device tree bindings are also updated
for this changes.

Signed-off-by: Chirag Parekh <chirag.parekh@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../bindings/pinctrl/xlnx,zynqmp-pinctrl.txt       |   14 +++++----
 drivers/pinctrl/pinctrl-zynqmp.c                   |   29 +++++++++++++++++--
 2 files changed, 34 insertions(+), 9 deletions(-)

diff --git a/Documentation/devicetree/bindings/pinctrl/xlnx,zynqmp-pinctrl.txt b/Documentation/devicetree/bindings/pinctrl/xlnx,zynqmp-pinctrl.txt
index a565e55..c74223c 100644
--- a/Documentation/devicetree/bindings/pinctrl/xlnx,zynqmp-pinctrl.txt
+++ b/Documentation/devicetree/bindings/pinctrl/xlnx,zynqmp-pinctrl.txt
@@ -52,7 +52,9 @@ to specify in a pinconf subnode:
   sdio0_0_grp - sdio0_2_grp, sdio0_0_pc_grp - sdio0_2_pc_grp,
   sdio0_0_cd_grp - sdio0_2_cd_grp, sdio0_0_wp_grp - sdio0_2_wp_grp,
   sdio1_0_grp - sdio1_1_grp, sdio1_0_pc_grp - sdio1_1_pc_grp,
-  sdio1_0_cd_grp - sdio1_1_cd_grp, sdio1_0_wp_grp - sdio1_2_wp_grp, nand0_0_grp,
+  sdio1_0_cd_grp - sdio1_1_cd_grp, sdio1_0_wp_grp - sdio1_2_wp_grp,
+  nand0_0_grp, nand0_0_ce_grp - nand0_1_ce_grp,
+  nand0_0_rb_grp - nand0_1_rb_grp, nand0_0_dqs_grp - nand0_1_dqs_grp,
   can0_0_grp - can0_18_grp, can1_0_grp - can1_19_grp,
   uart0_0_grp - uart0_18_grp, uart1_0_grp - uart1_18_grp,
   i2c0_0_grp - i2c0_18_grp, i2c1_0_grp - i2c1_19_grp,
@@ -75,10 +77,10 @@ to specify in a pinconf subnode:
    ethernet0, ethernet1, ethernet2, ethernet3, gemtsu0, usb0, usb1, mdio0,
    mdio1, mdio2, mdio3, qspi0, qspi_fbclk, qspi_ss, spi0, spi1, spi0_ss,
    spi1_ss, sdio0, sdio0_pc, sdio0_wp, sdio0_cd, sdio1, sdio1_pc, sdio1_wp,
-   sdio1_cd, nand0, can0, can1, uart0, uart1, i2c0, i2c1, ttc0_clk, ttc0_wav,
-   ttc1_clk, ttc1_wav, ttc2_clk, ttc2_wav, ttc3_clk, ttc3_wav, swdt0_clk,
-   swdt0_rst, swdt1_clk, swdt1_rst, gpio0, pmu0, pcie0, csu0, dpaux0, pjtag0,
-   trace0, trace0_clk, testscan0
+   sdio1_cd, nand0, nand0_ce, nand0_rb, nand0_dqs, can0, can1, uart0, uart1,
+   i2c0, i2c1, ttc0_clk, ttc0_wav, ttc1_clk, ttc1_wav, ttc2_clk, ttc2_wav,
+   ttc3_clk, ttc3_wav, swdt0_clk, swdt0_rst, swdt1_clk, swdt1_rst, gpio0, pmu0,
+   pcie0, csu0, dpaux0, pjtag0, trace0, trace0_clk, testscan0
 
 The following driver-specific properties as defined here are valid to specify in
 a pin configuration subnode:
@@ -112,7 +114,7 @@ Example:
 			conf-tx {
 				pins = "MIO19";
 				bias-disable;
-				schmitt-cmos = <PIN_INPUT_TYPE_CMOS>
+				schmitt-cmos = <PIN_INPUT_TYPE_CMOS>;
 			};
 		};
 	};
diff --git a/drivers/pinctrl/pinctrl-zynqmp.c b/drivers/pinctrl/pinctrl-zynqmp.c
index 840021b..d1cafd9 100644
--- a/drivers/pinctrl/pinctrl-zynqmp.c
+++ b/drivers/pinctrl/pinctrl-zynqmp.c
@@ -104,6 +104,9 @@ enum zynqmp_pinmux_functions {
 	ZYNQMP_PMUX_sdio1_cd,
 	ZYNQMP_PMUX_sdio1_wp,
 	ZYNQMP_PMUX_nand0,
+	ZYNQMP_PMUX_nand0_ce,
+	ZYNQMP_PMUX_nand0_rb,
+	ZYNQMP_PMUX_nand0_dqs,
 	ZYNQMP_PMUX_ttc0_clk,
 	ZYNQMP_PMUX_ttc0_wav,
 	ZYNQMP_PMUX_ttc1_clk,
@@ -310,9 +313,14 @@ static const unsigned int sdio1_1_pc_pins[] = {70};
 static const unsigned int sdio1_1_cd_pins[] = {77};
 static const unsigned int sdio1_1_wp_pins[] = {69};
 
-static const unsigned int nand0_0_pins[] = {9, 10, 11, 12, 13, 14, 15, 16, 17,
-						 18, 19, 20, 21, 22, 23, 24,
-						 25, 26, 27, 28, 32};
+static const unsigned int nand0_0_pins[] = {13, 14, 15, 16, 17, 18, 19, 20,
+						21, 22, 23, 24, 25};
+static const unsigned int nand0_0_ce_pins[] = {9};
+static const unsigned int nand0_0_rb_pins[] = {10, 11};
+static const unsigned int nand0_0_dqs_pins[] = {12};
+static const unsigned int nand0_1_ce_pins[] = {26};
+static const unsigned int nand0_1_rb_pins[] = {27, 28};
+static const unsigned int nand0_1_dqs_pins[] = {32};
 
 static const unsigned int can0_0_pins[] = {2, 3};
 static const unsigned int can0_1_pins[] = {6, 7};
@@ -787,6 +795,12 @@ static const struct zynqmp_pctrl_group zynqmp_pctrl_groups[] = {
 	DEFINE_ZYNQMP_PINCTRL_GRP(sdio1_1_cd),
 	DEFINE_ZYNQMP_PINCTRL_GRP(sdio1_1_wp),
 	DEFINE_ZYNQMP_PINCTRL_GRP(nand0_0),
+	DEFINE_ZYNQMP_PINCTRL_GRP(nand0_0_ce),
+	DEFINE_ZYNQMP_PINCTRL_GRP(nand0_0_rb),
+	DEFINE_ZYNQMP_PINCTRL_GRP(nand0_0_dqs),
+	DEFINE_ZYNQMP_PINCTRL_GRP(nand0_1_ce),
+	DEFINE_ZYNQMP_PINCTRL_GRP(nand0_1_rb),
+	DEFINE_ZYNQMP_PINCTRL_GRP(nand0_1_dqs),
 	DEFINE_ZYNQMP_PINCTRL_GRP(can0_0),
 	DEFINE_ZYNQMP_PINCTRL_GRP(can0_1),
 	DEFINE_ZYNQMP_PINCTRL_GRP(can0_2),
@@ -1206,6 +1220,12 @@ static const char * const sdio1_wp_groups[] = {"sdio1_0_wp_grp",
 		"sdio1_1_wp_grp"};
 
 static const char * const nand0_groups[] = {"nand0_0_grp"};
+static const char * const nand0_ce_groups[] = {"nand0_0_ce_grp",
+						"nand0_1_ce_grp"};
+static const char * const nand0_rb_groups[] = {"nand0_0_rb_grp",
+						"nand0_1_rb_grp"};
+static const char * const nand0_dqs_groups[] = {"nand0_0_dqs_grp",
+						"nand0_1_dqs_grp"};
 
 static const char * const can0_groups[] = {"can0_0_grp", "can0_1_grp",
 		"can0_2_grp", "can0_3_grp", "can0_4_grp", "can0_5_grp",
@@ -1401,6 +1421,9 @@ static const struct zynqmp_pinmux_function zynqmp_pmux_functions[] = {
 	DEFINE_ZYNQMP_PINMUX_FUNCTION(sdio1_wp, 0x08),
 	DEFINE_ZYNQMP_PINMUX_FUNCTION(sdio1_cd, 0x08),
 	DEFINE_ZYNQMP_PINMUX_FUNCTION(nand0, 0x02),
+	DEFINE_ZYNQMP_PINMUX_FUNCTION(nand0_ce, 0x02),
+	DEFINE_ZYNQMP_PINMUX_FUNCTION(nand0_rb, 0x02),
+	DEFINE_ZYNQMP_PINMUX_FUNCTION(nand0_dqs, 0x02),
 	DEFINE_ZYNQMP_PINMUX_FUNCTION(can0, 0x10),
 	DEFINE_ZYNQMP_PINMUX_FUNCTION(can1, 0x10),
 	DEFINE_ZYNQMP_PINMUX_FUNCTION(uart0, 0x60),
-- 
1.7.5.4

